config/configure_file.py

```diff
@@ -0,0 +1,219 @@
+#!/usr/bin/env python3
+
+# for the sake of portability, we try to avoid fstrings
+
+import argparse
+import os
+import re
+import string
+import sys
+
+_MAX_VARNAME_SIZE = 256
+
+def is_valid_varname(s, start = None, stop = None):
+    return re.fullmatch("\w+", s[slice(start, stop)]) is not None
+
+def process_line(line_num, line, out_f, variable_map, used_variable_set):
```

Comment:
Unless I've missed out on some edge cases you're worried about, I think this function can be implemented much more simply with regular expressions. Here is a code snippet that does the main bit. I've used f-strings and left out the `used_variables_set` just for simplicity of demonstration. Anyway, something like:

```
line = "Hello, I am @name, and you are @yourname."
varmap = {"name": "me", "yourname": "you"}

# this matches three groups
# 1. beginning of line or a white-space character
# 2. an alpha-numeric sequence appearing after a "@"
# 3. end of line or (white-space, period, comma)
my_reg = re.compile(r"(^|\s)\@([a-zA-Z0-9]+)($|[\s\.\,])")

for match in my_reg.findall(line):
    find = f"{match[0]}@{match[1]}{match[2]}"
    val = varmap.pop(match[1])
    replace = f"{match[0]}{val}{match[2]}"
    line = re.sub(find, replace, line, count=1)
```

Reply:
So on complication is that the variables are enclosed by `@` symbols on both sides . Consequently, we would match 
```
line = "Hello, I am @name@, and you are @yourname@."
```

With that said, I think it would probably be faster to use regex expressions. (When I was originally writing this function, I was intentionally trying to avoid some regex stuff in case we decided we wanted to transcribe the function into C or bash -- but that's a moot point)